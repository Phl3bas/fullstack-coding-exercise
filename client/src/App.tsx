import * as React from "react";
import { Filter } from "./components/filter";
import { UserTable } from "./components/userTable";

function App() {
  const [filterValue, setFilterValue] = React.useState<string>("");

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setFilterValue(e.target.value);
  };

  return (
    <div>
      <h1>Super Amasing User Search!</h1>
      <Filter handleChange={handleChange} />
      <UserTable>hi</UserTable>
    </div>
  );
}

export default App;
