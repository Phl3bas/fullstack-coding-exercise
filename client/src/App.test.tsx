import * as React from "react";
import {
  render,
  waitFor,
  screen,
  fireEvent,
  waitForElementToBeRemoved,
  queryByTestId,
} from "@testing-library/react";
import App from "./App";
import { enableFetchMocks } from "jest-fetch-mock";
import "@testing-library/jest-dom";

enableFetchMocks();

beforeEach(() => {
  fetchMock.resetMocks();
  fetchMock.mockResponse(
    JSON.stringify([
      {
        name: "Joe Clarke",
        email: "Email@email.com",
        location: "bournemouth",
      },
      {
        name: "Bob Bobbington",
        email: "thebobbinator@gmail.com",
        location: "Manchester",
      },
      {
        name: "Lizzie Clarke",
        email: "lizzierascal@gmail.com",
        location: "Poole",
      },
    ])
  );
});

test("to match snapshot", async () => {
  const { container } = await waitFor(() => render(<App />));
  expect(container).toMatchSnapshot();
});

test("render without crashing", async () => {
  await waitFor(() => render(<App />));
});

test("should render a filter input", async () => {
  const { getByTestId } = await waitFor(() => render(<App />));

  const input = getByTestId("filter");

  expect(input).toBeInTheDocument();
});

test("should render a table of results", async () => {
  const { getByTestId, getAllByTestId, debug } = await waitFor(() =>
    render(<App />)
  );
  const table = getByTestId("table");
  expect(table).toBeInTheDocument();

  const tableRows = getAllByTestId("table-row");

  expect(tableRows.length).toBe(3);
});

test("should filter by firstname and/or lastname", async () => {
  const { getByTestId, getAllByTestId, container } = await waitFor(() =>
    render(<App />)
  );

  expect(getAllByTestId("table-row").length).toBe(3);

  fireEvent.change(
    getByTestId("filter").querySelector("input") as HTMLInputElement,
    {
      target: {
        value: "J",
      },
    }
  );

  expect(getAllByTestId("table-row").length).toBe(1);
});

test("should display a message if no results from filter", async () => {
  const { debug, getAllByTestId, getByTestId, container } = await waitFor(() =>
    render(<App />)
  );

  expect(getAllByTestId("table-row").length).toBe(3);

  fireEvent.change(
    getByTestId("filter").querySelector("input") as HTMLInputElement,
    {
      target: {
        value: "z",
      },
    }
  );

  expect(getByTestId("alert")).toBeInTheDocument();
});

export {};
