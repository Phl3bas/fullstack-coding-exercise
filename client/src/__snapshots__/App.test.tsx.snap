// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`to match snapshot 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <h1
      class="makeStyles-header-1"
    >
      User Search
    </h1>
    <form
      class="makeStyles-form-3"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-filter-4"
        data-testid="filter"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
          data-shrink="false"
          for="outlined-basic"
          id="outlined-basic-label"
          style="color: rgb(121, 134, 203);"
        >
          Filter users
        </label>
        <div
          class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
          style="color: rgb(121, 134, 203);"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiFilledInput-input"
            id="outlined-basic"
            type="text"
            value=""
          />
        </div>
      </div>
    </form>
    <div
      class="MuiPaper-root makeStyles-paper-7 MuiPaper-elevation2 MuiPaper-rounded"
    >
      <div
        class="MuiPaper-root MuiTableContainer-root makeStyles-container-5 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <table
          aria-label="user-table"
          class="MuiTable-root makeStyles-table-6 MuiTable-stickyHeader"
          data-testid="table"
        >
          <thead
            class="MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head makeStyles-head-8 MuiTableCell-sizeSmall MuiTableCell-stickyHeader"
                scope="col"
              >
                Name
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head makeStyles-head-8 MuiTableCell-sizeSmall MuiTableCell-stickyHeader"
                scope="col"
              >
                Email
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head makeStyles-head-8 MuiTableCell-sizeSmall MuiTableCell-stickyHeader"
                scope="col"
              >
                Location
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-hover"
              data-testid="table-row"
            >
              <th
                class="MuiTableCell-root MuiTableCell-body makeStyles-cell-9 MuiTableCell-sizeSmall"
                role="cell"
                scope="row"
              >
                Joe Clarke
              </th>
              <td
                class="MuiTableCell-root MuiTableCell-body makeStyles-cell-9 MuiTableCell-sizeSmall"
              >
                Email@email.com
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body makeStyles-cell-9 MuiTableCell-sizeSmall"
              >
                bournemouth
              </td>
            </tr>
            <tr
              class="MuiTableRow-root MuiTableRow-hover"
              data-testid="table-row"
            >
              <th
                class="MuiTableCell-root MuiTableCell-body makeStyles-cell-9 MuiTableCell-sizeSmall"
                role="cell"
                scope="row"
              >
                Bob Bobbington
              </th>
              <td
                class="MuiTableCell-root MuiTableCell-body makeStyles-cell-9 MuiTableCell-sizeSmall"
              >
                thebobbinator@gmail.com
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body makeStyles-cell-9 MuiTableCell-sizeSmall"
              >
                Manchester
              </td>
            </tr>
            <tr
              class="MuiTableRow-root MuiTableRow-hover"
              data-testid="table-row"
            >
              <th
                class="MuiTableCell-root MuiTableCell-body makeStyles-cell-9 MuiTableCell-sizeSmall"
                role="cell"
                scope="row"
              >
                Lizzie Clarke
              </th>
              <td
                class="MuiTableCell-root MuiTableCell-body makeStyles-cell-9 MuiTableCell-sizeSmall"
              >
                lizzierascal@gmail.com
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body makeStyles-cell-9 MuiTableCell-sizeSmall"
              >
                Poole
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
